### DISCUSSION TITLE
Draggable UI

### DISCUSSION THREAD

**Discussion by GoldcatGames** [ScratchAddons/ScratchAddons#4577] [Ideas]
What if you can edit the positions of the UI by dragging it around in the addon description/settings?

**Comment by WorldLanguages** [comment]
What do you mean by "edit the positions of the UI"?

**Reply by GoldcatGames** [reply]
concept:
![Screenshot 2022-05-15 141949](https://user-images.githubusercontent.com/87254177/168495579-c82fe44c-3c07-4b23-996f-ea2575d8543b.png)
basically dragging parts of the UI in the editor and moving them around. (also yes I used the scratch addons editor darkmode description preview to show it.)

**Reply by WorldLanguages** [reply]
Ohh, you meant the Scratch editor UI.

**Reply by GoldcatGames** [reply]
yeah

**Reply by GoldcatGames** [reply]
Is There Any Vaildation That This Will Be A Feature Or No

**Comment by mxmou** [comment]
If it had the option to remove elements as well, for example by dragging them into a trash can, this could replace editor-stage-left, tutorials-button, and #4475.

**Reply by GoldcatGames** [reply]
that's a good idea too, the only thing is where would the trash-can be?

**Reply by Norbiros** [reply]
Also, how i can get items from trash? If you deleete something and you want to add it again.

**Reply by mxmou** [reply]
Something like this might work:
![image](https://user-images.githubusercontent.com/51849865/168675881-a80f892e-fe0f-428f-af9d-ed3711f2f10e.png)

**Reply by Norbiros** [reply]
Okay. But items inside it, should be just icon of some kind, maybe text or normal element (without resizing)?

**Reply by mxmou** [reply]
I think there is enough space to show the entire element.

**Reply by Norbiros** [reply]
Okay, so another question, elements have different sizes. So how we can make them look good in one element (trash)?
![sometasdinasd](https://user-images.githubusercontent.com/89837724/168677676-a8b6c982-ebe6-47d0-8fc9-e6e8d9881923.png)
Put them in order from left > right?

**Reply by GoldcatGames** [reply]
prob just snap into spots dedicated to the dragged in element?

**Reply by mxmou** [reply]
They could be arranged horizontally or vertically.

**Reply by GoldcatGames** [reply]
true

**Comment by Norbiros** [comment]
Maybe we should split page as this:
![soething](https://user-images.githubusercontent.com/89837724/168599908-ba681a58-7242-4edd-80e7-4639665b4db5.png)
Right element is group of 5 other elements, you can change order of them inside light green element, but you can't move small elements to red box, but can move whole right element to left.
*You can also have light green, and red elements on the same site*

**Reply by GoldcatGames** [reply]
that's a good idea

**Reply by Secret-chest** [reply]
What about the stage controls? also what about `columns`?

**Reply by Norbiros** [reply]
How we can change something that is horizontal, to be vertical and look good?

**Reply by mxmou** [reply]
I don't think anything inside the Code, Costumes or Sounds tab should be controlled by this addon. Only the general layout.

**Reply by GoldcatGames** [reply]
stage controls would be grouped with the stage, it wouldn't give you the option to put it vertically, and also what about moving the sprites side tab or the sounds side tab

**Reply by Secret-chest** [reply]
> stage controls would be grouped with the stage, it wouldn't give you the option to put it vertically, and also what about moving the sprites side tab or the sounds side tab

yes, but what if you want them to be flipped, like scratch 2?

**Reply by mxmou** [reply]
> > stage controls would be grouped with the stage, it wouldn't give you the option to put it vertically, and also what about moving the sprites side tab or the sounds side tab
> 
> yes, but what if you want them to be flipped, like scratch 2?

I think it would make sense to allow reordering them. It should also allow moving menu bar items.

**Reply by GoldcatGames** [reply]
> > > stage controls would be grouped with the stage, it wouldn't give you the option to put it vertically, and also what about moving the sprites side tab or the sounds side tab
> > 
> > 
> > yes, but what if you want them to be flipped, like scratch 2?
> 
> I think it would make sense to allow reordering them. It should also allow moving menu bar items.

ok

**Comment by Cool-Scratcher** [comment]
I really like this idea! Maybe there could also be resizable areas (when you hover over the edge a resize cursor appears, then you could resize it to whatever you wanted). I suggested this idea a long time ago in #2217.

The sizing and positions of the UI should save. Combined with the drag idea it would allow users to _fully_ customize the user layout to whatever they like.

I'd imagine it would open up a lot of bugs in other addons, though.

**Reply by GoldcatGames** [reply]
if this gets implemented as a feature of the addon, there has to be a cap on how big you can make it

**Comment by Oren-Lindsey** [comment]
this is a good idea but sounds like a terror to implement,

**Reply by TheColaber** [reply]
a preview for editor dark mode was a terror to implement so...

**Reply by Oren-Lindsey** [reply]
yeah that's true I guess

**Reply by GoldcatGames** [reply]
there's no gain without any pain ;-;
